### ISSUE TITLE
_merge_chunk_array could allocate memory more carefully

### ISSUE THREAD

**Issue by rabernat** [zarr-developers/zarr-python#2038]
I was reviewing the Pipeline and has some observations. A key function is `_merge_chunk_array`

https://github.com/zarr-developers/zarr-python/blob/b8baa6868c8fa95e6b2948c9fd9c725331ca23ec/src/zarr/codecs/pipeline.py#L298-L332

There are two issues with this implementation

1. `if existing_chunk_array is None: ` this code path always allocates a new empty array and then copies data from the original selection into it? Why? This seems like a great opportunity to avoid and extra memory copy. Can't we just use a view of the original data (`values`) when appropriate? This happens e.g. when writing to chunks which are smaller than the selection
2. The same code path always creates data of shape `chunk_spec.shape`, _even for the last chunk of an array_, which might be much smaller than the chunk shape.

An example which highlights both of these inefficiencies is as follows:

```python
import numpy as np
import zarr
zarr.array(np.ones(101), chunks=(100,))
```

**Comment by rabernat** [comment]
Related: this `is_total_slice` check will generally fail for the last chunk because the last chunk selection will not be the same size as `chunk_spec.shape`.

https://github.com/zarr-developers/zarr-python/blob/b8baa6868c8fa95e6b2948c9fd9c725331ca23ec/src/zarr/codecs/pipeline.py#L361-L362

When writing, this will trigger unnecessary trips to the store to read chunks, even though the data will be completely overwritten.

**Comment by d-v-b** [comment]
good catch @rabernat, I don't know this code very well but have a few high-level additions
- `is_total_slice` has been a source of [problems / confusion in the past](https://github.com/zarr-developers/zarr-python/issues/757). I suspect we need aggressive testing of this routine.
- I wish we had performance tooling that could _show me_ the memory / IO load associated with your examples. There must be something in the python space we could use for this.

**Comment by rabernat** [comment]
[Tracemalloc](https://docs.python.org/3/library/tracemalloc.html) should allow some tracing of allocations.

A good target would be the following: if I create a memorystore with uncompressed data, I should be able to get this data all the way to a numpy array without copying any bytes.

**Comment by dcherian** [comment]
Ive used the memory_profiler ipython extension to good effect. More recently scalene is getting a lot of attention

**Comment by d-v-b** [comment]
maybe it would make sense to create a specific hatch environment with useful benchmarking / performance monitoring tools, so we could do stuff like 

`hatch run perf:<invoke memory monitoring tool>`

**Comment by d-v-b** [comment]
points in this issue were addressed by #3885, #2784, and the addition of our benchmark suite, so I'm closing this as completed.
